src/runtime-rs/crates/resource/src/network/network_with_netns.rs

```diff
@@ -0,0 +1,216 @@
+        }
+        Ok(interfaces)
+    }
+
+    async fn routes(&self) -> Result<Vec<agent::Route>> {
+        let inner = self.inner.read().await;
+        let mut routes = vec![];
+        for e in &inner.entity_list {
+            let mut list = e.network_info.routes().await.context("routes")?;
+            routes.append(&mut list);
+        }
+        Ok(routes)
+    }
+
+    async fn neighs(&self) -> Result<Vec<agent::ARPNeighbor>> {
+        let inner = self.inner.read().await;
+        let mut neighs = vec![];
+        for e in &inner.entity_list {
+            let mut list = e.network_info.neighs().await.context("neighs")?;
+            neighs.append(&mut list);
+        }
+        Ok(neighs)
+    }
+}
+
+async fn get_entity_from_netns(config: &NetworkWithNetNsConfig) -> Result<Vec<NetworkEntity>> {
+    info!(
+        sl!(),
+        "get network entity for config {:?} tid {:?}",
+        config,
+        nix::unistd::gettid()
+    );
+    let mut entity_list = vec![];
+    let _netns_guard = netns::NetnsGuard::new(&config.netns_path)
+        .context("net netns guard")
+        .unwrap();
+    let (connection, handle, _) = rtnetlink::new_connection().context("new connection")?;
+    let thread_handler = tokio::spawn(connection);
+    defer!({
+        thread_handler.abort();
+    });
+
+    let mut links = handle.link().get().execute();
+
+    let idx = AtomicU32::new(0);
+    while let Some(link) = links.try_next().await? {
+        let link = link::get_link_from_message(link);
+        let attrs = link.attrs();
+
+        if (attrs.flags & libc::IFF_LOOPBACK as u32) != 0 {
+            continue;
+        }
+
+        let idx = idx.fetch_add(1, Ordering::Relaxed);
+        let (endpoint, network_info) = create_endpoint(&handle, link.as_ref(), idx, config)
+            .await
+            .context("create endpoint")?;
+
+        entity_list.push(NetworkEntity::new(endpoint, network_info));
+    }
+
+    Ok(entity_list)
+}
+
+async fn create_endpoint(
+    handle: &rtnetlink::Handle,
+    link: &dyn link::Link,
+    idx: u32,
+    config: &NetworkWithNetNsConfig,
+) -> Result<(Arc<dyn Endpoint>, Arc<dyn NetworkInfo>)> {
+    let _netns_guard = netns::NetnsGuard::new(&config.netns_path)
+        .context("net netns guard")
+        .unwrap();
+    let attrs = link.attrs();
+    let link_type = link.r#type();
+    let endpoint: Arc<dyn Endpoint> = if is_physical_iface(&attrs.name)? {
+        info!(
+            sl!(),
+            "physical network interface found: {} {:?}",
+            &attrs.name,
+            nix::unistd::gettid()
+        );
+        let t = PhysicalEndpoint::new(&attrs.name, &attrs.hardware_addr)
+            .context("new physical endpoint")?;
+        Arc::new(t)
+    } else {
+        info!(
+            sl!(),
+            "{} network interface found: {}", &link_type, &attrs.name
+        );
+        // check if this is a dummy interface which has a vhost-user socket associated with it
```

Comment:
Is there any code related to `vhost-user`?

Reply:
Thanks for your review, currently no related code, I'll delete the comment.